ludwig/models/abstractmodel.py

```diff
@@ -0,0 +1,242 @@
+import logging
+from abc import ABCMeta, abstractmethod
+from collections import OrderedDict
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+import numpy as np
+import torch
+
+from ludwig.combiners.combiners import Combiner
+from ludwig.constants import COMBINED, LOSS, NAME, TIED, TYPE
+from ludwig.features.base_feature import InputFeature, OutputFeature
+from ludwig.features.feature_registries import input_type_registry, output_type_registry
+from ludwig.utils.algorithms_utils import topological_sort_feature_dependencies
+from ludwig.utils.metric_utils import get_scalar_from_ludwig_metric
+from ludwig.utils.misc_utils import get_from_registry
+from ludwig.utils.torch_utils import LudwigModule, reg_loss
+
+logger = logging.getLogger(__name__)
+
+
+class AbstractModel(LudwigModule, metaclass=ABCMeta):
+    def __init__(self, random_seed=None):
+        self._random_seed = random_seed
+
+        # TODO: with change to misc_utils.set_random_seed() this may be redundant
+        #       seems to be required for test_api.py::test_api_training_determinism
+        if random_seed is not None:
+            torch.random.manual_seed(random_seed)
+
+        super().__init__()
+
+    @classmethod
+    def build_inputs(cls, input_features_def: List[Dict[str, Any]]) -> Dict[str, InputFeature]:
+        """Builds and returns input features in topological order."""
+        input_features = OrderedDict()
+        input_features_def = topological_sort_feature_dependencies(input_features_def)
+        for input_feature_def in input_features_def:
+            input_features[input_feature_def[NAME]] = cls.build_single_input(input_feature_def, input_features)
+        return input_features
+
+    @staticmethod
+    def build_single_input(
+        input_feature_def: Dict[str, Any], other_input_features: Dict[str, InputFeature]
+    ) -> InputFeature:
+        """Builds a single input feature from the input feature definition."""
+        logger.debug(f"Input {input_feature_def[TYPE]} feature {input_feature_def[NAME]}")
+
+        encoder_obj = None
+        if input_feature_def.get(TIED, None) is not None:
+            tied_input_feature_name = input_feature_def[TIED]
+            if tied_input_feature_name in other_input_features:
+                encoder_obj = other_input_features[tied_input_feature_name].encoder_obj
+
+        input_feature_class = get_from_registry(input_feature_def[TYPE], input_type_registry)
+        input_feature_obj = input_feature_class(input_feature_def, encoder_obj)
+
+        return input_feature_obj
+
+    @classmethod
+    def build_outputs(cls, output_features_def: List[Dict[str, Any]], combiner: Combiner) -> Dict[str, OutputFeature]:
```

Comment:
Could this be a `@staticmethod`? It looks like there's a call to `build_single_output`, but this is already a `@staticmethod`.

Reply:
This is a classmethod, so that it can be overwritten in subclasses (e.g. GBM).